src/instructlab/sdg/utils/chunking.py

```diff
@@ -36,15 +53,82 @@ def chunk_document(documents: List, server_ctx_size, chunk_word_count) -> List[s
-    text_splitter = RecursiveCharacterTextSplitter(
-        separators=["\n\n", "\n", " "],
+    chunk_size = _num_chars_from_tokens(no_tokens_per_doc)
+    chunk_overlap = _DEFAULT_CHUNK_OVERLAP
+    text_splitter = RecursiveCharacterTextSplitter.from_language(
+        language=Language.MARKDOWN,
         chunk_size=_num_chars_from_tokens(no_tokens_per_doc),
         chunk_overlap=_DEFAULT_CHUNK_OVERLAP,
     )
 
+    # Determine file type for heuristics, default with markdown
     for docs in documents:
-        temp = text_splitter.create_documents([docs])
-        content.extend([item.page_content for item in temp])
+        # Try Except Block if Magika fails
+        try:
+            m = Magika()
+            docs_bytes = docs.encode("utf-8")
+            res = m.identify_bytes(docs_bytes)
+            file_type = res.output.ct_label.lower()
 
+            # Full list of supported languages of Langchain
+            # supported_types = [e.value for e in Language]
+            # Checks for file types:
+            file_type_to_language = {
+                "go": Language.GO,
+                "java": Language.JAVA,
+                "javascript": Language.JS,  # Adjusted to match the previous variable name Language.js
+                "php": Language.PHP,
+                "python": Language.PYTHON,
+                "ruby": Language.RUBY,
+                "rust": Language.RUST,
+                "latex": Language.LATEX,
+                "html": Language.HTML,
+                "cs": Language.CSHARP,
+                "c": Language.C,
+                "perl": Language.PERL,
+            }
+            if file_type in file_type_to_language:
+                language = file_type_to_language[file_type]
+                text_splitter = RecursiveCharacterTextSplitter.from_language(
+                    language=language,
+                    chunk_size=chunk_size,
+                    chunk_overlap=chunk_overlap,
+                )
+                temp = text_splitter.create_documents([docs])
+                content.extend([item.page_content for item in temp])
+
+            # Default case set for markdown, we assume most cases: pdf->md
+            else:
+                # Falls back to default case
+                text_splitter = RecursiveCharacterTextSplitter.from_language(
+                    language=Language.MARKDOWN,
+                    chunk_size=chunk_size,
+                    chunk_overlap=chunk_overlap,
+                )
+                # Use regex to remove unnecessary dashes in front of pipe characters in a markdown table.
+                docs = re.sub(r"-{2,}\|", "-|", docs)
+                # Remove unnecessary spaces in front of pipe characters in a markdown table.
+                docs = re.sub(r"\  +\|", " |", docs)
+                temp = text_splitter.create_documents([docs])
+                content.extend([item.page_content for item in temp])
+
+        except UnicodeDecodeError as e:
+            for docs in documents:
```

Comment:
I would be careful about how we're handling these exceptions. For instance, if `UnicodeDecodeError` was thrown, then it's possible we weren't able to parse the documents correctly when performing language detection. 

If the intention here is to fallback to the default case, then we would need to apply the same logic as in the `else` block above. 

My overall recommendation is to either cut down on the scope of our try/catch here, or to move the `content.extend` logic to happen within a `finally` block.

Reply:
You are absolutely correct. I would say ValueError, superclass of UnicodeDecodeError shouldn't be thrown into the try except logic. However, UnicodeDecodeError is a common error when Magika, the file type light weight detection model, fails to take a proper input; thus catching it. I've made amends to the logger for proper error messages.